ddspipe_core/src/cpp/communication/dds/DdsBridge.cpp

```diff
@@ -175,6 +97,178 @@ void DdsBridge::disable() noexcept
     }
 }
 
+void DdsBridge::create_all_tracks()
+{
+    const auto& ids = participants_->get_participants_ids();
+
+    // Figure out what writers need to be created
+    std::set<ParticipantId> writers_to_create;
+
+    for (const ParticipantId& id : ids)
+    {
+        const auto& routes_it = routes_.find(id);
+
+        if (routes_it != routes_.end())
+        {
+            // The reader has a route. Create only the writers in the route.
+            const auto& writers_ids = routes_it->second;
+            writers_to_create.insert(writers_ids.begin(), writers_ids.end());
+        }
+        else
+        {
+            // The reader doesn't have a route. Create every writer (+ itself if repeater)
+            auto writers_ids = ids;
+
+            if (!participants_->get_participant(id)->is_repeater())
+            {
+                // The participant is not a repeater. Do not add its writer.
+                writers_ids.erase(id);
+            }
+
+            writers_to_create.insert(writers_ids.begin(), writers_ids.end());
+        }
+    }
+
+    // Create the writers.
+    std::map<ParticipantId, std::shared_ptr<IWriter>> writers;
+
+    for (const auto& id : writers_to_create)
+    {
+        std::shared_ptr<IParticipant> participant = participants_->get_participant(id);
+        writers[id] = participant->create_writer(*topic_);
+    }
+
+    // Add the writers to the tracks they have routes for.
+    add_writers_to_tracks_(writers);
+}
+
+void DdsBridge::add_to_tracks(
+        const ParticipantId& discoverer_participant_id)
+{
+    // A new subscriber has been discovered.
+    // Check if the writer for this participant already exists.
+    for (const auto& tracks_it : tracks_)
+    {
+        const auto& track = tracks_it.second;
+
+        if (track->has_writer(discoverer_participant_id))
+        {
+            // The writer already exists. There is nothing to do. Exit.
+            return;
+        }
+    }
+
+    std::shared_ptr<IParticipant> participant = participants_->get_participant(discoverer_participant_id);
+
+    // Create the writer.
+    std::map<ParticipantId, std::shared_ptr<IWriter>> writer;
+    writer[discoverer_participant_id] = participant->create_writer(*topic_);
+
+    // Add the writer to the tracks it has routes for.
+    add_writers_to_tracks_(writer);
```

Comment:
Couldn't there be an `add_writer_to_tracks_` ?

Reply:
Yes, there could. I didn't write it because the code would almost be identical to the code in add_writers_to_tracks_. Do you consider it worth it to write it?

Reply:
You added it already :)